core/src/main/scala/io/paradoxical/aetr/core/db/dao/StepDb.scala

```diff
@@ -105,25 +133,94 @@ class StepDb @Inject()(
     provider.withDB(update).map(updated => updated == 1)
   }
 
-  def getPendingRuns(): Future[List[Run]] = {
-    val pendingRoots = runs.query.filter(r => r.state === RunState.Pending && (r.id === r.root)).result
+  def findUnlockedRuns(state: RunState): Future[List[RootId]] = {
+    val pendingRoots = runs.query.filter(r =>
+      r.state === state &&
+      r.id === r.root &&
+      (r.actionLockedTill.isEmpty || r.actionLockedTill <= Instant.now(clock))
+    ).result
 
-    provider.withDB(pendingRoots).flatMap(roots => {
-      Future.sequence(roots.map(r => getRun(Root(r.id.value)))).map(_.toList)
+    provider.withDB(pendingRoots).map(roots => {
+      roots.map(r => RootId(r.id.value)).toList
+    })
+  }
+
+  /**
+   * Locks all nodes in a tree via an optimistic lock on the tree nodes
+   * if the lock is acquired, the action is run, then the lock is released
+   *
+   * @param rootId
+   * @param block
+   * @tparam T
+   * @return
+   */
+  def lock[T](rootId: RootId)(block: Run => T): Future[Option[T]] = {
+    val now = Instant.now(clock)
+
+    val newLockId = LockId(UUID.randomUUID())
+
+    val lockExpirationTime = now.plus(config.dbLockTime.toSeconds, ChronoUnit.SECONDS)
+
+    // NOTE: Only locking on the root id, since you can't lock on sub nodes anyways
+    // this means that the remaining tree will look "unlocked" from a db perspective
+    // but it also means we dont have to update N nodes, we only have to update 1.
+    //
+    // Flow is:
+    // if nobody's ever locked it,
+    // or the lock is expired (it was set to expire and it is currently later than that)
+    // then acquire a lock
+
+    val lockQuery = runs.query.
+      filter(r =>
+        r.id === rootId.asRunInstance &&
+        r.actionLockedTill.isEmpty || r.actionLockedTill <= now
+      ).
+      map(r => (r.actionLockedTill, r.lockId)).
+      update((Some(lockExpirationTime), Some(newLockId)))
+
+    // unlock the instance and the lock key is the
+    // time we expected to lock till
+    val unlockQuery = runs.query.
+      filter(r =>
+        r.id === rootId.asRunInstance &&
+        r.lockId === newLockId
+      ).
+      map(r => (r.actionLockedTill, r.lockId)).
+      update((None, None))
+
+    provider.withDB(lockQuery).flatMap(results => {
+      // if the lock was acquired, resolve the tree
+      // and allow someone to do work with it
+      // and attempt a safe unlock
+      if (results > 0) {
+        logger.debug(s"Acquired lock $newLockId to expire on $lockExpirationTime")
+
+        for {
+          data <- getRun(rootId).map(block).map(Some(_))
```

Comment:
What happens if this fails?

Reply:
If the block or get or really anything fails, the lock is orphaned and times out. Check the test where I injected an adjustable clock and simulated this scenario